libs/payments/cart/src/lib/checkout.service.ts

```diff
@@ -2,43 +2,131 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import { Injectable } from '@nestjs/common';
+import { EligibilityService } from '@fxa/payments/eligibility';
+import { PayPalManager, PaypalCustomerManager } from '@fxa/payments/paypal';
 import {
   StripeClient,
   StripeManager,
   StripeSubscription,
+  SubplatInterval,
+  TaxAddress,
 } from '@fxa/payments/stripe';
-import { PayPalManager, PaypalCustomerManager } from '@fxa/payments/paypal';
-import { Injectable } from '@nestjs/common';
-import { ResultCart } from './cart.types';
 import { CheckoutError, CheckoutPaymentError } from './checkout.error';
+import {
+  CartTotalMismatchError,
+  CartEligibilityMismatchError,
+  CartEmailNotFoundError,
+} from './cart.error';
+import { CartManager } from './cart.manager';
+import { ResultCart } from './cart.types';
+import { handleEligibilityStatusMap } from './cart.utils';
+import { ContentfulService } from '@fxa/shared/contentful';
+import { AccountManager } from '@fxa/shared/account/account';
 
 @Injectable()
 export class CheckoutService {
   constructor(
     private stripeClient: StripeClient,
     private stripeManager: StripeManager,
     private paypalCustomerManager: PaypalCustomerManager,
-    private paypalManager: PayPalManager
+    private paypalManager: PayPalManager,
+    private cartManager: CartManager,
+    private eligibilityService: EligibilityService,
+    private contentfulService: ContentfulService,
+    private accountManager: AccountManager
   ) {}
 
-  async prePaySteps(cart: ResultCart) {
-    // TODO:
-    // - If uid not present, create user stub account and update cart
-    // - Check if customer exists, create via stripe manager and update cart
-    // - Make sure shipping address of customer is updated with cart taxAddress
-    // - Check if customer is eligible for product in cart
-    // - Check if amount to be charged matches amount in cart
-    // - Cancel incomplete subscriptions to price user is subscribing to
-
-    const { stripeCustomerId } = cart;
-    if (!stripeCustomerId) {
-      // TODO: this is a placeholder and won't be necessary once the above todo items are done
-      throw new Error("Stripe customer doesn't exist");
+  async prePaySteps(cart: ResultCart, locale: string) {
+    let customer, stripeCustomerId, uid;
+    const taxAddress = cart.taxAddress as any as TaxAddress;
+
+    if (!cart.email) {
+      throw new CartEmailNotFoundError(cart.id);
     }
-    const customer = await this.stripeManager.fetchActiveCustomer(
+
+    // if stripeCustomerId not found, create stub stripe account
+    if (!cart.stripeCustomerId) {
+      customer = await this.stripeManager.createPlainCustomer({
+        email: cart.email,
+        taxAddress: taxAddress,
+      });
+
+      stripeCustomerId = customer.id;
+    } else {
+      stripeCustomerId = cart.stripeCustomerId;
+
+      customer = await this.stripeManager.fetchActiveCustomer(stripeCustomerId);
+    }
```

Comment:
Doesn't customer creation need to happen second, so that we can set the uid on the stripe customer?

Reply:
(gets set as the `userid` metadata on the stripe customer)